Qt example doesn't compile

Here's output of `make -j8` of step "2 Build the verovio-qt library":

```                                             
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++1y  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.10 -Wall -W -Wno-unused-parameter -fPIC -DVEROVIOQT_LIBRARY -DQT_DEPRECATED_WARNINGS -DQT_NO_DEBUG -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I../lib -I. -I../lib/include -I../../../include -I../../../include/vrv -I../../../include/pugi -I../../../include/utf8 -I../../../libmei -I/Users/**********/Qt/5.9.9/clang_64/lib/QtQuick.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtGui.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtQml.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtNetwork.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtCore.framework/Headers -I. -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/AGL.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/mkspecs/macx-clang -F/Users/**********/Qt/5.9.9/clang_64/lib -o scenegraphdevicecontext.o ../lib/src/scenegraphdevicecontext.cpp
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++1y  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.10 -Wall -W -Wno-unused-parameter -fPIC -DVEROVIOQT_LIBRARY -DQT_DEPRECATED_WARNINGS -DQT_NO_DEBUG -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I../lib -I. -I../lib/include -I../../../include -I../../../include/vrv -I../../../include/pugi -I../../../include/utf8 -I../../../libmei -I/Users/**********/Qt/5.9.9/clang_64/lib/QtQuick.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtGui.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtQml.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtNetwork.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtCore.framework/Headers -I. -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/AGL.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/mkspecs/macx-clang -F/Users/**********/Qt/5.9.9/clang_64/lib -o view.o ../lib/src/view.cpp
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++1y  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.10 -Wall -W -Wno-unused-parameter -fPIC -DVEROVIOQT_LIBRARY -DQT_DEPRECATED_WARNINGS -DQT_NO_DEBUG -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I../lib -I. -I../lib/include -I../../../include -I../../../include/vrv -I../../../include/pugi -I../../../include/utf8 -I../../../libmei -I/Users/**********/Qt/5.9.9/clang_64/lib/QtQuick.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtGui.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtQml.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtNetwork.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/lib/QtCore.framework/Headers -I. -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/AGL.framework/Headers -I/Users/**********/Qt/5.9.9/clang_64/mkspecs/macx-clang -F/Users/**********/Qt/5.9.9/clang_64/lib -o moc_view.o moc_view.cpp
In file included from In file included from ../lib/src/view.cpp:moc_view.cpp8::
10In file included from :
../lib/include/vrvqt/view.hIn file included from :./../lib/include/vrvqt/view.h18::
18:
../lib/include/vrvqt/scenegraphdevicecontext.h./../lib/include/vrvqt/scenegraphdevicecontext.h::9797::8383::  errorerror: : non-virtualnon-virtual  membermember  functionfunction  markedmarked  'override''override'  hideshides  virtualvirtual 
member      
member       functionfunction

    void DrawRoundedRectangle(int x, int y, int width, int height, double radius) override;    void DrawRoundedRectangle(int x, int y, int width, int height, double radius) override;

                                                                                  ^                                                                                  ^

../../../include/vrv/devicecontext.h:149:18: ../../../include/vrv/devicecontext.h:149note:: 18:hidden  overloaded notevirtual:  functionhidden
       overloaded'vrv::DeviceContext::DrawRoundedRectangle'  virtualdeclared  functionhere: 
type       'vrv::DeviceContext::DrawRoundedRectangle'mismatch  declaredat  here:5th  typeparameter  mismatch('int' vs 'double') at
 5th parameter ('int' vs 'double')
    virtual void DrawRoundedRectangle(int x, int y, int width, int height, int radius) = 0;
In file included from                  ^../lib/src/scenegraphdevicecontext.cpp
:8:
../lib/include/vrvqt/scenegraphdevicecontext.h:97:83: error: non-virtual    virtual void DrawRoundedRectangle(int x, int y, int width, int height, int radius) = 0; 
member function                 ^ 
marked 'override' hides virtual member
      function
    void DrawRoundedRectangle(int x, int y, int width, int height, double radius) override;
                                                                                  ^
../../../include/vrv/devicecontextIn file included from .h:moc_view.cpp149::1018:
: In file included from ./../lib/include/vrvqt/view.h:note18: hidden overloaded virtual function
      'vrv::DeviceContext::DrawRoundedRectangle':
 declared here:./../lib/include/vrvqt/scenegraphdevicecontext.h :type119 :mismatch81 :at  5th errorparameter:  ('int' vs 'double')non-virtual
 member function marked 'override' hides virtual
      member function
In file included from     virtual void DrawRoundedRectangle(int x, int y, int width, int height, int radius) = 0;../lib/src/view.cpp
:8                 ^:

In file included from ../lib/include/vrvqt/view.h:18:
../lib/include/vrvqt/scenegraphdevicecontext.h:119:81    void StartGraphic(vrv::Object *object, std::string gClass, std::string gId) override;:
                                                                                 ^
error: non-virtual member function../../../include/vrv/devicecontext.h :marked 202'override': 18hides:  virtual
      notemember:  functionhidden 
overloaded virtual function 'vrv::DeviceContext::StartGraphic'
      declared here: different number of parameters (5 vs 3)
    void StartGraphic(vrv::Object *object, std::string gClass, std::string gId) override;    virtual void StartGraphic(

                                                                                ^                 ^

../../../include/vrv/devicecontext.h:202:18: note: hidden overloaded virtual function 'vrv::DeviceContext::StartGraphic'
      declared here: different number of parameters (5 vs 3)
    virtual void StartGraphic(
                 ^
In file included from ../lib/src/scenegraphdevicecontext.cpp:8:
../lib/include/vrvqt/scenegraphdevicecontext.h:119:81: error: non-virtual member function marked 'override' hides virtual
      member function
    void StartGraphic(vrv::Object *object, std::string gClass, std::string gId) override;
                                                                                ^
../../../include/vrv/devicecontext.h:202:18: note: hidden overloaded virtual function 'vrv::DeviceContext::StartGraphic'
      declared here: different number of parameters (5 vs 3)
    virtual void StartGraphic(
                 ^
moc_view.cpp:131:1: error: unknown type name 'QT_INIT_METAOBJECT'
QT_INIT_METAOBJECT const QMetaObject vrvQt::View::staticMetaObject = { {
^
moc_view.cpp:131:37: error: expected ';' after top level declarator
QT_INIT_METAOBJECT const QMetaObject vrvQt::View::staticMetaObject = { {
                                    ^
                                    ;
4 errors generated.
make: *** [moc_view.o] Error 1
make: *** Waiting for unfinished jobs....
2 errors generated.
make: *** [scenegraphdevicecontext.o] Error 1
In file included from ../lib/src/view.cpp:8:
In file included from ../lib/include/vrvqt/view.h:13:
In file included from /Users/**********/Qt/5.9.9/clang_64/lib/QtQuick.framework/Headers/QQuickItem:1:
In file included from /Users/**********/Qt/5.9.9/clang_64/lib/QtQuick.framework/Headers/qquickitem.h:43:
In file included from /Users/**********/Qt/5.9.9/clang_64/lib/QtQuick.framework/Headers/qtquickglobal.h:43:
In file included from /Users/**********/Qt/5.9.9/clang_64/lib/QtQml.framework/Headers/qtqmlglobal.h:47:
In file included from /Users/**********/Qt/5.9.9/clang_64/lib/QtCore.framework/Headers/qglobal.h:109:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:643:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3131:32: error: 
      allocating an object of abstract class type 'vrvQt::SceneGraphDeviceContext'
    return unique_ptr<_Tp>(new _Tp(_VSTD::forward<_Args>(__args)...));
                               ^
../lib/src/view.cpp:63:34: note: in instantiation of function template specialization
      'std::__1::make_unique<vrvQt::SceneGraphDeviceContext, vrvQt::View *, QSGNode *&>' requested here
        m_qtDeviceContext = std::make_unique<SceneGraphDeviceContext>(this, node);
                                 ^
../../../include/vrv/devicecontext.h:139:18: note: unimplemented pure virtual method 'DrawSimpleBezierPath' in
      'SceneGraphDeviceContext'
    virtual void DrawSimpleBezierPath(Point bezier[4]) = 0;
                 ^
../../../include/vrv/devicecontext.h:149:18: note: unimplemented pure virtual method 'DrawRoundedRectangle' in
      'SceneGraphDeviceContext'
    virtual void DrawRoundedRectangle(int x, int y, int width, int height, int radius) = 0;
                 ^
../../../include/vrv/devicecontext.h:179:18: note: unimplemented pure virtual method 'MoveTextVerticallyTo' in
      'SceneGraphDeviceContext'
    virtual void MoveTextVerticallyTo(int y) = 0;
                 ^
../../../include/vrv/devicecontext.h:202:18: note: unimplemented pure virtual method 'StartGraphic' in
      'SceneGraphDeviceContext'
    virtual void StartGraphic(
                 ^
3 errors generated.
make: *** [view.o] Error 1
```

--- a/bindings/qt/README.md
+++ b/bindings/qt/README.md
@@ -8,7 +8,7 @@ Building the demo for Qt involves following steps:
 ```
 mkdir VEROVIO_ROOT/tools/build-library
 cd VEROVIO_ROOT/tools/build-library
-cmake -DBUILD_AS_LIBRARY=ON ..
+cmake -DBUILD_AS_LIBRARY=ON ../../cmake
 make -j8
 ln -s $(pwd)/libverovio.so ..
 ```

--- a/bindings/qt/demo/Options.qml
+++ b/bindings/qt/demo/Options.qml
@@ -16,7 +16,7 @@ Item {
 
     width: layout.width
 
-    property string fileName: ":/mei/test.mei"
+    property string fileUrl: ":/mei/test.mei"
     property string musicFontName: "Leipzig"
 
     Column {
@@ -37,7 +37,7 @@ Item {
             Row {
                 spacing: 5
                 TextField {
-                    text: root.fileName
+                    text: root.fileUrl
                     Layout.fillWidth: true
                     Layout.minimumWidth: grid.minimumInputSize
                     Layout.alignment: Qt.AlignLeft | Qt.AlignBaseline
@@ -54,7 +54,7 @@ Item {
                     title: "Please choose a file"
                     folder: shortcuts.home
                     onAccepted: {
-                        root.fileName = fileDialog.fileUrl
+                        root.fileUrl = fileDialog.fileUrl
                     }
                 }
             }
@@ -79,7 +79,7 @@ Item {
     }
 
     Settings {
-        property alias fileName: root.fileName
+        property alias fileUrl: root.fileUrl
         property alias musicFontName: root.musicFontName
     }
 }

--- a/bindings/qt/demo/main.qml
+++ b/bindings/qt/demo/main.qml
@@ -51,7 +51,7 @@ Item {
 
         resourcesDataPath: "../../../data"
 
-        fileName: options.fileName
+        fileName: options.fileUrl
         musicFontName: options.musicFontName
         musicFontPath: {
             if (options.musicFontName == "Bravura")
@@ -118,7 +118,7 @@ Item {
         keys: ["text/plain"]
         onDropped: {
             if (drop.hasText) {
-                options.fileName = drop.text.trim()
+                options.fileUrl = drop.text.trim()
             }
         }
     }

--- a/bindings/qt/demo/verovio-qt-demo.pro
+++ b/bindings/qt/demo/verovio-qt-demo.pro
@@ -26,10 +26,9 @@ else: unix:!android: target.path = /opt/$${TARGET}/bin
     INCLUDEPATH += ../../../include/pugi
     INCLUDEPATH += ../../../include/utf8
     INCLUDEPATH += ../../../libmei
+    INCLUDEPATH += ../lib/include
+
+    LIBS += -L../build-library -lverovio-qt
     LIBS += -L../../../tools -lverovio
 }
 
-{
-    INCLUDEPATH += ../../qt/lib/include
-    LIBS += -L../../qt/build-library -lverovio-qt
-}

--- a/bindings/qt/lib/include/vrvqt/scenegraphdevicecontext.h
+++ b/bindings/qt/lib/include/vrvqt/scenegraphdevicecontext.h
@@ -38,7 +38,7 @@ public:
      */
     ///@{
     SceneGraphDeviceContext(QQuickItem *quickItem, QSGNode *node);
-    virtual ~SceneGraphDeviceContext() = default;
+    ~SceneGraphDeviceContext() override = default;
     vrv::ClassId GetClassId() const override { return vrv::CUSTOM_DEVICE_CONTEXT; }
     void Clear();
     ///@}
@@ -57,6 +57,7 @@ public:
     ///@{
     QList<QSGGeometryNode *> GetGeometryNodesForId(QString id);
     QList<TextQuickItem *> GetQuickItemsForId(QString id);
+    QList<TextQuickItem *> FindQuickItemsByType(QString type);
     QStringList GetIdsForQuickItem(QQuickItem *item);
     ///@}
 
@@ -85,6 +86,7 @@ public:
      * @name Drawing methods
      */
     ///@{
+    void DrawSimpleBezierPath(vrv::Point bezier[4]) override;
     void DrawComplexBezierPath(vrv::Point bezier1[4], vrv::Point bezier2[4]) override;
     void DrawCircle(int x, int y, int radius) override;
     void DrawEllipse(int x, int y, int width, int height) override;
@@ -94,7 +96,7 @@ public:
         int n, vrv::Point points[], int xoffset, int yoffset, int fill_style = vrv::AxODDEVEN_RULE) override;
     void DrawRectangle(int x, int y, int width, int height) override;
     void DrawRotatedText(const std::string &text, int x, int y, double angle) override;
-    void DrawRoundedRectangle(int x, int y, int width, int height, double radius) override;
+    void DrawRoundedRectangle(int x, int y, int width, int height, int radius) override;
     void DrawText(
         const std::string &text, const std::wstring wtext = L"", int x = VRV_UNSET, int y = VRV_UNSET) override;
     void DrawMusicText(const std::wstring &text, int x, int y, bool setSmuflGlyph) override;
@@ -110,13 +112,14 @@ public:
     void StartText(int x, int y, vrv::data_HORIZONTALALIGNMENT alignment = vrv::HORIZONTALALIGNMENT_left) override;
     void EndText() override;
     void MoveTextTo(int x, int y, vrv::data_HORIZONTALALIGNMENT alignment) override;
+    void MoveTextVerticallyTo(int y) override;
     ///@}
 
     /**
      * @name Method for starting, restarting and ending a graphic
      */
     ///@{
-    void StartGraphic(vrv::Object *object, std::string gClass, std::string gId) override;
+    void StartGraphic(vrv::Object *object, std::string gClass, std::string gId, bool primary, bool prepend) override;
     void EndGraphic(vrv::Object *object, vrv::View *view) override;
     void ResumeGraphic(vrv::Object *object, std::string gId) override;
     void EndResumedGraphic(vrv::Object *object, vrv::View *view) override;
@@ -160,15 +163,16 @@ private:
      * @name Functions used to translate from the logical to device coordinates.
      */
     ///@{
-    template <class T> float translate(T x)
+    float translate(float x)
     {
         // User-scale: Everything has to be scaled, thus it is not possible to differentiate between UserScaleX and
         // UserScaleY. If there is no requirement in other DeviceContexts to have different user-scales it makes sense
         // to merge the two values to a single user-scale.
-        return x / DEFINITION_FACTOR * GetUserScaleX();
+        float userScale = static_cast<float>(GetUserScaleX());
+        return x * userScale / DEFINITION_FACTOR;
     }
-    template <class T> float translateX(T x) { return translate(x + m_logicalOrigin.x); }
-    template <class T> float translateY(T y) { return translate(y + m_logicalOrigin.y); }
+    float translateX(float x) { return translate(x + m_logicalOrigin.x); }
+    float translateY(float y) { return translate(y + m_logicalOrigin.y); }
     ///@}
 
 private:

--- a/bindings/qt/lib/include/vrvqt/textquickitem.h
+++ b/bindings/qt/lib/include/vrvqt/textquickitem.h
@@ -56,8 +56,8 @@ public:
     void setColor(QColor color);
 
     /**
-      * Returns if the text item is empty, i.e. it has no text elements.
-      */
+     * Returns if the text item is empty, i.e. it has no text elements.
+     */
     bool isEmpty();
 
     /**

--- a/bindings/qt/lib/include/vrvqt/toolkit.h
+++ b/bindings/qt/lib/include/vrvqt/toolkit.h
@@ -36,6 +36,7 @@ class Toolkit : public QObject {
     Q_PROPERTY(QString musicFontPath MEMBER m_musicFontPath WRITE setMusicFontPath)
     Q_PROPERTY(QString verovioTextFontPath MEMBER m_verovioTextFontPath WRITE setVerovioTextFontPath)
     Q_PROPERTY(QString breaks WRITE setBreaks READ getBreaks)
+    Q_PROPERTY(QString transpose WRITE setTranspose READ getTranspose)
     Q_PROPERTY(int pageCount MEMBER m_pageCount READ pageCount NOTIFY pageCountChanged)
     Q_PROPERTY(int displayWidth MEMBER m_displayWidth WRITE setDisplayWidth)
     Q_PROPERTY(int displayHeight MEMBER m_displayHeight WRITE setDisplayHeight)
@@ -87,6 +88,7 @@ public:
     int getSpacingSystem() const;
     int getSpacingStaff() const;
     QString getBreaks() const;
+    QString getTranspose() const;
     ///@}
 
 public slots:
@@ -101,6 +103,7 @@ public slots:
     void setMusicFontPath(QString musicFontPath);
     void setVerovioTextFontPath(QString verovioTextFontPath);
     void setBreaks(QString breaks);
+    void setTranspose(QString transpose);
     void setDisplayWidth(int displayWidth);
     void setDisplayHeight(int displayHeight);
     void setScale(int scale);
@@ -205,6 +208,9 @@ private:
 
     // path to the resource dir
     QString m_resourcesDataPath;
+
+    // transposition string
+    QString m_transpose;
 
     QString m_fileName;
     QString m_fileContent;

--- a/bindings/qt/lib/include/vrvqt/view.h
+++ b/bindings/qt/lib/include/vrvqt/view.h
@@ -59,6 +59,12 @@ signals:
      */
     void toolkitChanged(Toolkit *toolkit);
 
+    /**
+     * @name This signal is used internally to request a new height of the view from the render thread. On some
+     * platforms (e.g. Android), setHeight has to be called on the GUI thread.
+     */
+    void requestNewHeight(int height);
+
 protected:
     /**
      * @name Reimplemented function from QQuickItem.

--- a/bindings/qt/lib/src/scenegraphdevicecontext.cpp
+++ b/bindings/qt/lib/src/scenegraphdevicecontext.cpp
@@ -16,6 +16,7 @@
 #include <QSGGeometryNode>
 #include <QSGSimpleRectNode>
 #include <QScreen>
+#include <QThread>
 #include <cmath>
 
 //----------------------------------------------------------------------------
@@ -64,6 +65,10 @@ void SceneGraphDeviceContext::AddGeometryNode(QSGGeometryNode *node)
 
 void SceneGraphDeviceContext::AddQuickItem(TextQuickItem *item)
 {
+    // The scene graph rendering may be done on a separate thread (i.e. not the GUI thread), thus we have to ensure that
+    // the new item is on the same thread as the m_quickItem.
+    item->moveToThread(m_quickItem->thread());
+
     item->setParentItem(m_quickItem); // visual parent
     item->setParent(m_quickItem); // object parent (for proper cleanup)
 
@@ -80,6 +85,17 @@ QList<QSGGeometryNode *> SceneGraphDeviceContext::GetGeometryNodesForId(QString 
 QList<TextQuickItem *> SceneGraphDeviceContext::GetQuickItemsForId(QString id)
 {
     return m_id2QuickItemMapping[id];
+}
+
+QList<TextQuickItem *> SceneGraphDeviceContext::FindQuickItemsByType(QString type)
+{
+    QList<TextQuickItem *> quickItems;
+    for (auto iter = m_id2QuickItemMapping.begin(); iter != m_id2QuickItemMapping.end(); ++iter) {
+        if (iter.key().startsWith(type)) {
+            quickItems << iter.value();
+        }
+    }
+    return quickItems;
 }
 
 QStringList SceneGraphDeviceContext::GetIdsForQuickItem(QQuickItem *item)
@@ -130,6 +146,21 @@ vrv::Point SceneGraphDeviceContext::GetLogicalOrigin()
     return m_logicalOrigin;
 }
 
+void SceneGraphDeviceContext::DrawSimpleBezierPath(vrv::Point bezier[])
+{
+    qWarning() << "Warning:" << __FUNCTION__ << "not supported";
+}
+
+QRgb GetRgbFromPen(const vrv::Pen &pen)
+{
+    if (pen.GetColour() == AxNONE) {
+        return static_cast<QRgb>(AxBLACK);
+    }
+    else {
+        return static_cast<QRgb>(pen.GetColour());
+    }
+}
+
 void SceneGraphDeviceContext::DrawComplexBezierPath(vrv::Point bezier1[4], vrv::Point bezier2[4])
 {
     // Note: No support for vertex antialiasing. Use a top-level QQuickView with multisample antialiasing.
@@ -149,7 +180,7 @@ void SceneGraphDeviceContext::DrawComplexBezierPath(vrv::Point bezier1[4], vrv::
     node->setFlag(QSGNode::OwnsGeometry);
 
     QSGFlatColorMaterial *material = new QSGFlatColorMaterial;
-    material->setColor(static_cast<QRgb>(currentPen.GetColour()));
+    material->setColor(GetRgbFromPen(currentPen));
     node->setMaterial(material);
     node->setFlag(QSGNode::OwnsMaterial);
 
@@ -203,7 +234,7 @@ void SceneGraphDeviceContext::DrawCircle(int x, int y, int radius)
     node->setFlag(QSGNode::OwnsGeometry);
 
     QSGFlatColorMaterial *material = new QSGFlatColorMaterial;
-    material->setColor(static_cast<QRgb>(currentPen.GetColour()));
+    material->setColor(GetRgbFromPen(currentPen));
     node->setMaterial(material);
     node->setFlag(QSGNode::OwnsMaterial);
 
@@ -236,6 +267,29 @@ void SceneGraphDeviceContext::DrawLine(int x1, int y1, int x2, int y2)
 {
     vrv::Pen currentPen = m_penStack.top();
 
+    // Some OpenGL devices (e.g. on Android) do not allow to use GL_LINES with a line-width > 1. DrawLine with
+    // line-width > 1 is called e.g. for measure break or staff lines. For vertical and horizontal lines the DrawLine
+    // call can be easily translated into a DrawRectangle call. Diagonal lines could be implemented by using DrawPolygon
+    // but have not yet been observed.
+    if (translate(currentPen.GetWidth()) > 1) {
+        int x = std::min(x1, x2);
+        int y = std::min(y1, y2);
+        auto width = std::abs(x2 - x1);
+        auto height = std::abs(y2 - y1);
+
+        if (width == 0) {
+            DrawRectangle(x - (currentPen.GetWidth() / 2), y, width + currentPen.GetWidth(), height);
+            return;
+        }
+        else if (height == 0) {
+            DrawRectangle(x, y - (currentPen.GetWidth() / 2), width, height + currentPen.GetWidth());
+            return;
+        }
+        else {
+            qWarning() << "Warning: Calling DrawLine for a diagonal line with line-width > 1";
+        }
+    }
+
     QSGGeometry *geometry = new QSGGeometry(QSGGeometry::defaultAttributes_Point2D(), 2);
     geometry->setDrawingMode(GL_LINES);
     geometry->setLineWidth(translate(currentPen.GetWidth()));
@@ -243,7 +297,7 @@ void SceneGraphDeviceContext::DrawLine(int x1, int y1, int x2, int y2)
     geometry->vertexDataAsPoint2D()[1].set(translateX(x2), translateY(y2));
 
     QSGFlatColorMaterial *material = new QSGFlatColorMaterial;
-    material->setColor(static_cast<QRgb>(currentPen.GetColour()));
+    material->setColor(GetRgbFromPen(currentPen));
 
     QSGGeometryNode *node = new QSGGeometryNode;
     node->setGeometry(geometry);
@@ -272,7 +326,7 @@ void SceneGraphDeviceContext::DrawPolygon(int n, vrv::Point points[], int xoffse
     geometry->setDrawingMode(QSGGeometry::DrawTriangleStrip);
 
     QSGFlatColorMaterial *material = new QSGFlatColorMaterial;
-    material->setColor(static_cast<QRgb>(currentPen.GetColour()));
+    material->setColor(GetRgbFromPen(currentPen));
 
     // Reorder points so that they can be drawn with DrawTriangleStrip.
     int counter1 = 0;
@@ -302,9 +356,8 @@ void SceneGraphDeviceContext::DrawPolygon(int n, vrv::Point points[], int xoffse
 void SceneGraphDeviceContext::DrawRectangle(int x, int y, int width, int height)
 {
     vrv::Pen currentPen = m_penStack.top();
-
     QSGSimpleRectNode *node = new QSGSimpleRectNode;
-    node->setColor(static_cast<QRgb>(currentPen.GetColour()));
+    node->setColor(GetRgbFromPen(currentPen));
     qreal rectX = static_cast<qreal>(translateX(x));
     qreal rectY = static_cast<qreal>(translateY(y));
     qreal rectWidth = static_cast<qreal>(translate(width));
@@ -318,9 +371,9 @@ void SceneGraphDeviceContext::DrawRotatedText(const std::string &, int, int, dou
     // This function is also not implemented for SvgDeviceContext
 }
 
-void SceneGraphDeviceContext::DrawRoundedRectangle(int, int, int, int, double)
+void SceneGraphDeviceContext::DrawRoundedRectangle(int x, int y, int width, int height, int radius)
 {
-    qWarning() << "Warning:" << __FUNCTION__ << "not supported";
+    DrawRectangle(x, y, width, height);
 }
 
 void SceneGraphDeviceContext::StartText(int x, int y, vrv::data_HORIZONTALALIGNMENT alignment)
@@ -341,7 +394,14 @@ void SceneGraphDeviceContext::DrawText(const std::string &text, const std::wstri
         qWarning() << "Warning:" << __FUNCTION__ << "does not yet support specifying x and y";
     }
 
-    QFont font(QString::fromStdString(m_fontStack.top()->GetFaceName()));
+    // Verovio uses "Times" for layout calculation, thus we use it as our default. Note: Some texts (e.g. lyrics) don't
+    // have a face-name.
+    QString fontName = QString::fromStdString(m_fontStack.top()->GetFaceName());
+    if (fontName.isEmpty()) {
+        fontName = "Times";
+    }
+
+    QFont font(fontName);
 
     // Note: Verovio calls it point size but it is actually pixel size. Qt can only handle pixel size in int, thus it
     // would be better to use point size.
@@ -402,6 +462,11 @@ void SceneGraphDeviceContext::MoveTextTo(int x, int y, vrv::data_HORIZONTALALIGN
     }
 }
 
+void SceneGraphDeviceContext::MoveTextVerticallyTo(int y)
+{
+    MoveTextTo(0, y, vrv::HORIZONTALALIGNMENT_NONE);
+}
+
 void SceneGraphDeviceContext::SetTextPositionAndAlignment(int x, int y, vrv::data_HORIZONTALALIGNMENT alignment)
 {
     m_currentTextQuickItem->setX(static_cast<double>(translateX(x)));
@@ -454,8 +519,13 @@ void SceneGraphDeviceContext::DrawBackgroundImage(int, int)
     // This function is also not implemented for SvgDeviceContext
 }
 
-void SceneGraphDeviceContext::StartGraphic(vrv::Object *object, std::string, std::string gId)
+void SceneGraphDeviceContext::StartGraphic(
+    vrv::Object *object, std::string gClass, std::string gId, bool primary, bool prepend)
 {
+    if (prepend) {
+        qWarning() << "Warning:" << __FUNCTION__ << " with parameter prepend = true not supported";
+    }
+
     m_activeGraphicObjectsStack.push(ActiveGraphic(QString::fromStdString(gId), object));
 }
 

--- a/bindings/qt/lib/src/textquickitem.cpp
+++ b/bindings/qt/lib/src/textquickitem.cpp
@@ -11,6 +11,8 @@
 
 #include <QPainter>
 
+//#define DRAW_DEBUG_BOUNDING_BOXES
+
 namespace vrvQt {
 
 TextQuickItem::TextQuickItem()
@@ -29,10 +31,16 @@ void TextQuickItem::appendText(QString text, QFont font)
 
 void TextQuickItem::calcPos()
 {
+    // Note: It is not possible to re-use the existing bounding-box of the verovio-object because the stored
+    // values may be out-of-date. Thus we have to recalculate the correct position and baseline by examining the font.
+
     int textWidth = 0;
     int textAscent = 0;
     int textDescent = 0;
     int textXShift = 0;
+
+    bool hasBoldFont = false;
+    bool hasItalicFont = false;
 
     bool firstLoop = true;
     for (TextWithFont &textWithFont : m_textParts) {
@@ -41,23 +49,39 @@ void TextQuickItem::calcPos()
         // call - e.g 48,5 of 51 ms)! Subsequent calls are much faster, so Qt seems to already have a caching mechanism
         // in place. There is no significant difference between using tightBoundingRect, boundingRect or merely
         // height/width.
-        // TODO: use already existing font calculation from BBoxDeviceContext (which is currently not stored anywhere)
-        // for ascent, descent, height, width
         QFontMetrics fm(textWithFont.font);
 
-        auto boundingRect = fm.boundingRect(textWithFont.text);
+        // QFontMetrics::boundingRect(QChar) is more accurate than the QString overload, thus we prefer it if we only
+        // have 1 symbol (which is often the case, e.g. for music text). Using tightBoundingRect would result in many
+        // bounding boxes that are too narrow.
+        QRect boundingRect;
+        if (textWithFont.text.size() == 1)
+            boundingRect = fm.boundingRect(textWithFont.text.front());
+        else
+            boundingRect = fm.boundingRect(textWithFont.text);
 
         textWithFont.offset = textWidth;
-        textWidth += fm.width(textWithFont.text);
+        textWidth += boundingRect.right() - boundingRect.left();
         textAscent = std::max(textAscent, boundingRect.top() * -1);
         textDescent = std::max(textDescent, boundingRect.bottom());
         if (firstLoop) {
             // Note that the bounding rect can start at a negative x (e.g. italic font)
             textXShift = boundingRect.left();
-            textWidth -= textXShift;
+            if (textXShift < 0) textWidth -= textXShift;
             firstLoop = false;
         }
+
+        if (textWithFont.font.bold()) hasBoldFont = true;
+        if (textWithFont.font.italic()) hasItalicFont = true;
     }
+
+    // Security margin for bold/italic fonts because the bounding rects are sometimes too narrow
+    if (hasBoldFont || hasItalicFont) {
+        textWidth += 4;
+        textAscent += 4;
+        textDescent += 1;
+    }
+
     qreal newX = x() + textXShift;
     if (m_alignment == Qt::AlignRight) {
         setX(newX - textWidth);
@@ -70,8 +94,8 @@ void TextQuickItem::calcPos()
     }
     setY(y() - textAscent);
 
-    m_paintOffsetY += textAscent;
-    m_paintOffsetX += textXShift * -1;
+    m_paintOffsetY = textAscent;
+    m_paintOffsetX = -textXShift;
 
     int textHeight = textDescent + textAscent + 1;
 
@@ -98,5 +122,11 @@ void TextQuickItem::paint(QPainter *painter)
         painter->setFont(textWithFont.font);
         painter->drawText(m_paintOffsetX + textWithFont.offset, m_paintOffsetY, textWithFont.text);
     }
+
+#ifdef DRAW_DEBUG_BOUNDING_BOXES
+    painter->setPen(QColor(Qt::red));
+    painter->setBrush(QColor(255, 0, 0, 100));
+    painter->drawRect(0, 0, implicitWidth(), implicitHeight());
+#endif
 }
 } // namespace vrvQt

--- a/bindings/qt/lib/src/toolkit.cpp
+++ b/bindings/qt/lib/src/toolkit.cpp
@@ -10,8 +10,8 @@
 //----------------------------------------------------------------------------
 
 #include <QDebug>
+#include <QDir>
 #include <QFile>
-#include <QFileInfo>
 #include <QFontDatabase>
 #include <QUrl>
 
@@ -167,6 +167,15 @@ void Toolkit::setBreaks(QString breaks)
     }
 }
 
+void Toolkit::setTranspose(QString transpose)
+{
+    if (getTranspose() != transpose) {
+        m_verovioToolkit.SetOption("transpose", transpose.toStdString());
+        // "transpose" is used in LoadData
+        requestReloadData();
+    }
+}
+
 void Toolkit::setFileContent(QString fileContent)
 {
     if (m_fileContent != fileContent) {
@@ -230,6 +239,11 @@ int Toolkit::getSpacingStaff() const
 QString Toolkit::getBreaks() const
 {
     return QString::fromStdString(m_verovioToolkit.GetOption("spacingStaff"));
+}
+
+QString Toolkit::getTranspose() const
+{
+    return QString::fromStdString(m_verovioToolkit.GetOption("transpose"));
 }
 
 bool Toolkit::addFont(QString fontFilePath)

--- a/bindings/qt/lib/src/view.cpp
+++ b/bindings/qt/lib/src/view.cpp
@@ -21,6 +21,7 @@ namespace vrvQt {
 View::View()
 {
     setFlag(ItemHasContents, true);
+    connect(this, &View::requestNewHeight, this, &View::setHeight, Qt::QueuedConnection);
 }
 
 View::~View()
@@ -67,7 +68,7 @@ QSGNode *View::updatePaintNode(QSGNode *node, UpdatePaintNodeData *)
 
         node->removeAllChildNodes();
         for (auto child : childItems()) {
-            delete child;
+            child->deleteLater();
         }
         m_qtDeviceContext->Clear();
 
@@ -75,7 +76,10 @@ QSGNode *View::updatePaintNode(QSGNode *node, UpdatePaintNodeData *)
         m_verovioRenderingDirty = false;
 
         m_toolkit->setAdjustedDisplayHeightForPage(m_pageNumber, m_qtDeviceContext->GetScaledHeight());
-        setHeight(m_toolkit->adjustedDisplayHeightForPage(m_pageNumber));
+        auto newHeight = m_toolkit->adjustedDisplayHeightForPage(m_pageNumber);
+        if (height() != newHeight) {
+            emit requestNewHeight(newHeight);
+        }
     }
 
     return node;

--- a/bindings/qt/lib/verovio-qt.pro
+++ b/bindings/qt/lib/verovio-qt.pro
@@ -37,7 +37,7 @@ INCLUDEPATH += include
     INCLUDEPATH += ../../../include/pugi
     INCLUDEPATH += ../../../include/utf8
     INCLUDEPATH += ../../../libmei
-    LIBS += -L../../../tools -lverovio
+    LIBS += -L../../../tools/build-library -lverovio
 }
 
 unix {

